M701/M702: Set default Z value to 0

gudnimg opened the pull request:
Set default Z value to 0 to maintain backwards compatibility with older gcodes.


What needs to be tested:
* M600 with and without MMU
* See how print finish behavior is with MMU. Should now be like 3.12
* Any other functionality which uses M701 and M702.

Change in memory:
Flash: -24 bytes
SRAM: 0 bytes

gudnimg committed:
M701/M702: Set default Z value to 0

Set default Z value to 0 to maintain backwards compatibility with older gcodes.

gudnimg committed:
Merge branch 'prusa3d:MK3' into gudnimg-patch-1

gudnimg committed:
Restore old M701/M702 behavior

If the new Z parameter is not given,
assume that the old Z-lift behavior is desired.

Change in memory:
Flash: -28 bytes
SRAM: 0 bytes

gudnimg commented on Firmware/ultralcd.cpp:
```diff
@@ -3717,8 +3717,7 @@ static void lcd_wizard_load() {
         lcd_puts_at_P(0, 2, _T(MSG_LOADING_FILAMENT));
         loading_flag = true;
     }
-    gcode_M701(FILAMENTCHANGE_FIRSTFEED, 0);
```
Note I hope this change works OK. Otherwise we need another call to `raise_z_above(MIN_Z_FOR_LOAD)` here.

wavexx replied:
This seems to differ. "M701" will call gcode_M701(FILAMENTCHANGE_FIRSTFEED, 0xff) which seems to bypass ``MMU2::mmu2.load_filament_to_nozzle(mmuSlotIndex);``. Here we force slot 0 instead.
I can't test this at the moment to verify...

gudnimg replied:
You're correct!

I also noticed if slot `0xFF` is given with the MMU, then the M701 will act like the single color setup which is not good. I should probably split that MMU if statement into two nested if's even though it may take 4-8 bytes.
![image](https://user-images.githubusercontent.com/8218499/235798367-ca35b5a0-dc20-47f8-baf6-8af72510a7e0.png)

gudnimg replied:
Should hopefully be fixed now. Needs to be verified.

wavexx replied:
Looks good now.

It's almost a shame we have to put the if for all those raise_z_above().

During M701/M702 the Z parameter is always relative to the current position?

gudnimg replied:
> During M701/M702 the Z parameter is always relative to the current position?

Yes I believe that's correct. Because we use `raise_z` for the lift, which applies a delta to whatever the current position is. The delta is then also used to revert the Z move similar to how we unpark the XY axis in M600.

gudnimg committed:
Minor fix to M701 when MMU slot is out of range

If MMU slot is outside of range, then we rather the printer
not do anything. Currently it would accidentally assume
the MMU is disabled. This may have undesireable results.

No change in memory

gudnimg committed:
Select correct MMU slot in lcd_wizard_load

Change in memory:
Flash: +8 bytes
SRAM: 0 bytes

wavexx reviewed (approved):
Looks good now, but I couldn't test it on a real mmu
